Update badger2 dependency to v2.2007.2

gammazero opened the pull request:
This is needed to fix a protobuf conflict between badger and badger2 when both are linked into the same binary.

Issue described here:
https://github.com/dgraph-io/badger/issues/1208

Fix described here:
https://github.com/dgraph-io/badger/pull/1293

gammazero committed:
Update badger2 dependency to v2.2007.2

This is needed to fix a protobuf conflict between badger and badger2 when both are linked into the same binary.

aschmahmann commented on go.mod:
```diff
@@ -1,11 +1,20 @@
 
 require (
-	github.com/dgraph-io/badger/v2 v2.0.3
+	github.com/DataDog/zstd v1.4.5 // indirect
```
I generally try and keep indirect dependencies out of the go.mod file unless there's a good reason to keep them around. If you just update badger/v2 and then run `go mod tidy` things should clean up.

gammazero replied:
Done

aschmahmann commented on go.mod:
```diff
@@ -1,11 +1,20 @@
 require (
-	github.com/dgraph-io/badger/v2 v2.0.3
+	github.com/DataDog/zstd v1.4.5 // indirect
+	github.com/dgraph-io/badger/v2 v2.2007.2
```
CI is failing due to badger/v2 v2.2007.2 being unable to build on x86. https://discuss.dgraph.io/t/badger-build-failed-for-goos-linux-arm/9267

Try sending them a tiny PR for this that just makes the constant linked in that issue an `int64` and see what they think.

It looks like they may have other x86 problems (https://discuss.dgraph.io/t/fix-badger-builds-on-32-bit-machines/8679) but at least this won't block us on compiling.

gammazero committed:
remove indirect deps

gammazero committed:
Use latest badger2 to resolve protobuf conflict

This resolves a protobuf namespace conflict that would happen when both badger1 and badger2 are linked into the same binary.  For more information on this issue see: https://github.com/dgraph-io/badger/pull/1519

gammazero committed:
Update to badger with int overflow fix

gammazero committed:
Use correct default options that includes disabling mmap for the value log for 32bit

gammazero committed:
Drop ValueThreshold (inline size) back to default value.  This is what Lotus uses, and prevents transaction failures when writing larger batches with FileIO option used for the value log.

aschmahmann reviewed (approved):
Seems reasonable to me @Kubuxu is this OK by you?

gammazero commented:
@Kubuxu Here is an explanation of what was done to get badger2 working:

Depend on a newer version of badger2 that has two build fixes:
- Fixes integer overflow on 32bit builds
- Resolves protobuf namespace conflict with badger1

Use configuration options that includes disabling mmap for the value log and using FileIO instead.  This was apparently supposed to have been used previously, but was missed or was was not done because it resulted in transaction failures with large batch writes.  That issues is fixed by the following change:

Stop using LSMOnly options and use DefualtOptions instead. This changes the ValueThreshold to a much smaller size, and is the better choice because it fixes transaction failures for large batch writes, when used in conjunction with FileIO for the value log.  It is also better to use the default value as this is what Lotus uses.

In a future PR, batch writes will be handled by a BatchWriter instead of a Transaction, which should allow large batch writes to work with either IO configuration above.  Doing that will take care of https://github.com/ipfs/go-ds-badger/issues/92

Kubuxu reviewed (approved):
SGWM.

In Lotus, we are also using mmap IO. IDK how we feel about changing the default for that.

Kubuxu commented:
Also, we have ncreased LSM to the badger default, the 16MB was causing us to hit FD limits.

gammazero commented:
@Kubuxu If Lotus is using mmap IO, then I would not want to change that.  Using FileIO avoids mmap allocation failures during some 32bit tests, but I don't know that is something that lotus is concerned with.

Kubuxu commented:
We for sure use more than 3.5GiB memory, so no 32bit without pointer extensions for us.
